src/video/wayland/SDL_waylandwindow.c

```diff
@@ -2491,6 +2500,42 @@ void Wayland_SetWindowTitle(SDL_VideoDevice *_this, SDL_Window *window)
     }
 }
 
+int Wayland_SetWindowIcon(SDL_VideoDevice *_this, SDL_Window *window, SDL_Surface *icon)
+{
+    SDL_WindowData *wind = window->driverdata;
+    struct xdg_toplevel *toplevel = NULL;
+
+    if (!_this->driverdata->xdg_toplevel_icon_v1) {
+        return SDL_SetError("wayland: cannot set icon; xdg_toplevel_icon_v1 protocol not supported");
+    }
+
+    if (icon->w != icon->h) {
+        return SDL_SetError("wayland: icon width and height must be equal, got %ix%i", icon->w, icon->h);
+    }
+
+#ifdef HAVE_LIBDECOR_H
+    if (wind->shell_surface_type == WAYLAND_SURFACE_LIBDECOR && wind->shell_surface.libdecor.frame) {
+        toplevel = libdecor_frame_get_xdg_toplevel(wind->shell_surface.libdecor.frame);
+    } else
+#endif
+        if (wind->shell_surface_type == WAYLAND_SURFACE_XDG_TOPLEVEL && wind->shell_surface.xdg.roleobj.toplevel) {
+        toplevel = wind->shell_surface.xdg.roleobj.toplevel;
+    }
+
+    Wayland_ReleaseSHMBuffer(&wind->icon);
+    if (Wayland_AllocSHMBuffer(icon->w, icon->h, &wind->icon) != 0) {
+        return SDL_SetError("wayland: failed to allocate SHM buffer for the icon");
+    }
+
+    SDL_PremultiplyAlpha(icon->w, icon->h, icon->format->format, icon->pixels, icon->pitch, SDL_PIXELFORMAT_ARGB8888, wind->icon.shm_data, icon->w * 4);
+
+    if (toplevel) {
+        xdg_toplevel_icon_v1_set_icon_buffer(_this->driverdata->xdg_toplevel_icon_v1, toplevel, wind->icon.wl_buffer, 1);
```

Comment:
The spec allows for setting multiple icons at different sizes.

i.e:
set_icon_buffer(redSquare at 24px)
set_icon_buffer(redSquare at 48px)

The compositor might use the 48px icon might use the 24px one, it doesn't replace it.

if we wanted to change the icon to call, we would need to call:

set_icon_name("")
set_icon_buffer(blueSquare at 24px)

(or we can argue on the protocol for a spec change? it's not super intuitive, but it works)


In SDL we know nothing about the sizes given, so calling set_icon_name("") just before we set a buffer would be easiest.

Reply:
Ah, you're right. It's explicitly cleared now before setting a new icon.